### ISSUE TITLE
CascadedPolygonUnion very slow

### ISSUE THREAD

**Issue by rprimora** [NetTopologySuite/NetTopologySuite#306]
In my database I store all Australian postcode boundaries as GeoJSON. My user is allowed to add as much postcodes to its location as he likes. This forms sort of a "servicing area". It is how user defines which postcodes he is willing to work in.

Now, I allow my users to add the whole region. In this case I will add all postcodes that belong to this region. Also, we have a definition of cities where we sort of group postcodes so when user selects a city we automatically add postcodes that belong to this city. User can then proceede to fine tune this. So for instance, he can add Melbourne as a city and than remove 20-30 postcodes that he does not wish to service.

My problem arises when I do a merge of polygons. I want to merge all polygons so I can transfer this to clients browser where it will be displayed on google maps as an overlay. Sort of visualization of their servicing area.

On some 3000 postcodes this is extremely slow. My processor goes on 100%. I have a i7 8700K with 32Gb of RAM. All of my cores are on 100% and RAM shoots to over 1Gb during this time.

I use this method `CascadedPolygonUnion.Union(polygons)` as it shows best results but still...it takes good 5-8 seconds to complete this. Because of the nature of my application users can add and remove these postcodes as they like and I have to recalculate this polygone each time.

Am I doing something wrong? Is there room for improvement? This is my code...

```
var reader = new GeoJsonReader();
var polygons = new List<IGeometry>();

var count = 0;
foreach (var area in await ServicingLocation.GetPostcodeBoundariesForLocationAsync(LocationID.Value))
{
	if (area == "NULL")
		continue;
	count++;
	polygons.Add(reader.Read<Feature>(area).Geometry);
}

double distance = count * 0.00000007d;

var union = DouglasPeuckerSimplifier.Simplify(CascadedPolygonUnion.Union(polygons), distance);

/// Create the final feature by mergin all the polygones.
var finalFeature = new Feature() { Geometry = union };

/// Serialize the final feature and save it in location.
SetValue(() => GetDal<LocationDal>().LocationServicingArea, new GeoJsonWriter().Write(finalFeature));
```

I should also note that these are all tessellating polygons if that means anything to you.

**Comment by FObermaier** [comment]
You could simplify the input geometries instead of the result. 

What is the logic behind the computation of the simplification distance? IMHO this distance should be appropriate for the purpose.

**Comment by rprimora** [comment]
@FObermaier 

> You could simplify the input geometries instead of the result.

Simplifying the input geometries will result in a weird final geometry with gaps. Since these are postcode boundaries I need to retain the original boundary before merging and only after that I can simplify it. 

> What is the logic behind the computation of the simplification distance? IMHO this distance should be appropriate for the purpose.

I played with numbers until I got a satisfactory result. I do simplification so the final polygon has less points. This way it takes less space. If I don't do any of that than final polygon is, if I remember correctly, well over 5 Mb. This takes long time to tranfer to the client browser. This way it takes lot less space. Basically the more postcodes I have added the less detail I need.

**Comment by FObermaier** [comment]
> Simplifying the input geometries will result in a weird final geometry with gaps. Since these are postcode boundaries I need to retain the original boundary before merging and only after that I can simplify it.

It is easy to remove those artefacts from the result polygon and probably a lot cheaper.

**Comment by rprimora** [comment]
Can you elaborate this?

If I do a simplification prior to union I will get something like this.

![simplified prior to union](https://i.stack.imgur.com/y5A2j.jpg)

If I merge this I don't know if a gap is due to prior simplification of the polygon or if the user just didin't select a postcode there.

**Comment by FObermaier** [comment]
If you do a union, the overlaps should go away, the uncovered areas are tiny holes in your polygon.
You can filter those out by performing an `area > minArea` test and rebuild the polygon based on the shell and the remaining holes.

**Comment by rprimora** [comment]
This sounds promising. Can you give me an example of this as I am fairly green in this department?

**Comment by FObermaier** [comment]
_Beware, this has not seen a compiler_.

```C#
var reader = new GeoJsonReader();
var polygons = new List<IGeometry>();

int count = 0;
double distance = 0.000001d;
foreach (var area in await ServicingLocation.GetPostcodeBoundariesForLocationAsync(LocationID.Value))
{
	if (area == "NULL")
		continue;
	count++;
        var geom = reader.Read<Feature>(area).Geometry;
	polygons.Add(DouglasPeuckerSimplifier.Simplify(geom, distance));
}

var union = (IPolygon) CascadedPolygonUnion.Union(polygons);
var exteriorRing = (ILinearRing)union.ExteriorRing;
var interiorRings = new List<ILinearRing>(union.NumInteriorRings);
const double minArea = 0.0001 / ?? Is this reasonable ??
for (var i = 0; i < union.NumInteriorRings; i++) {
    var interiorRing = union.GetInterirorRingN(i);
    if (union.Factory.CreatePolygon(interiorRing).Area > minArea)
        interiorRings.Add(interiorRing);
}

union = union.Factory.CreatePolygon(exteriorRing, interiorRings.ToArray());
    
/// Create the final feature by mergin all the polygones.
var finalFeature = new Feature() { Geometry = union };

/// Serialize the final feature and save it in location.
SetValue(() => GetDal<LocationDal>().LocationServicingArea, new GeoJsonWriter().Write(finalFeature));

```

**Comment by rprimora** [comment]
Thank you for this fine example and it doesn't matter that it hasn'e been passed through compiler. It is still a good starting point for me.

At the moment I am bussy with something else but sometimes during the next week I will give it a go and see how it performs.

**Comment by airbreather** [comment]
> var union = (IPolygon) CascadedPolygonUnion.Union(polygons);

Watch out when casting to `IPolygon`... at work, our polygons have water areas carved out, so we often find islands within lakes within other land masses, which look kinda like this:
![image](https://user-images.githubusercontent.com/3792815/55479511-39b02a00-55ec-11e9-8605-eb8d2c20c8be.png)

That winds up being a `MULTIPOLYGON` with 2 polygons.

**Comment by rprimora** [comment]
@airbreather Thank you for pointing this out. What do you suggest, because I do have these examples. For instance, there are some Australian postcodes that cover huge amount of area and have holes in them for another small postcodes.

**Comment by FObermaier** [comment]
Iterate over each of these polygons and apply the same procedure.

**Comment by rprimora** [comment]
I have completely ditched this approach and delegated it all to the database. 

**Postcode**, **City** and **Region** all have a _geography_ column with area they cover. What this allows me to do is  do only one merge at any time which is faster than doing multiple merges. 

Another thing I did, since I use more optimized data type now, is always store the original polygon. Alongside this column that holds original polygon I have created another column which will hold my simplified GeoJSON that I will server to the user. This way I don't have to recalculate everything everytime I add or remove a postcode but I will have to recalculate simplified GeoJSON. I think it is a good tradeoff!

Here is an example:

A city exists that is consisted of 70 postcodes. Postcode areas have been merged into one polygon that is stored in the city. This polygon is C1.

1. User has no postcode.
2. User adds a postcode 3000(P1). Location shows area of P1
3. User adds a postcode 3005(P2). Location shows area P1+P2 = M1
4. User adds a postcode 3006(P3). Location shows area M1 + P3 = M2
5. ...
6. User adds a city(C1). Location shows area M2 + C1 = M3
